src/workerd/io/worker-fs.h

```diff
@@ -0,0 +1,192 @@
+    // the default root path is `/bundle`.
+    BUNDLE,
+    // The root for the built-in temp path.
+    TEMP,
+  };
+
+  // Gets the kj::Path mapped for the given known root.
+  kj::Path getPathForKnownRoot(KnownRoot root) const;
+
+  // Gets the base URL mapped for the given known root.
+  const jsg::Url& getUrlForKnownRoot(KnownRoot root) const;
+
+  // Returns the known root for the given path, if anything.
+  kj::Maybe<KnownRoot> tryGetKnownRootForUrl(const jsg::Url& path) const;
+
+  // Returns the known root for the given path, if anything.
+  kj::Maybe<KnownRoot> tryGetKnownRootForPath(kj::PathPtr path) const;
+
+  // Used to configure the FsMap instance, which is immutable once built.
+  class Builder final {
+   public:
+    Builder();
+
+    // The path must be parseable as a URL relative to the base URL "file:///".
+    // Only the path part of the URL is used.
+    void setPathForKnownRoot(KnownRoot root, kj::StringPtr path);
+    kj::Own<FsMap> finish();
+
+   private:
+    kj::Own<FsMap> map;
+  };
+
+  static kj::Own<FsMap> createDefault();
+
+ private:
+  kj::HashMap<KnownRoot, jsg::Url> paths;
+  friend class Builder;
+};
+
+// The directory delegate is used to provide the underlying implementation of
+// a directory in the WorkerFileSystem. It implements a subset of the kj::Directory
+// API focused on the specific methods that are needed for the WorkerFileSystem.
+// We currently have three DirectoryDelegate implementations:
+//   1. InMemoryDelegate: A virtual directory that is backed by an in-memory
+//      map. This is used for the interim directory structure when the FsMap
+//      if configured to use nested paths. For instance, if a known root path
+//      is mapped to be `/mybundle/modules`, then the `/mybundle` directory is
+//      backed by an InMemoryDelegate that contains the `/modules` directory.
+//   2. TempDirDelegate: A virtual directory that uses kj::InMemoryDirectory
+//      and kj::InMemoryFile to implement temporary file handling. The behavior
+//      of this directory is dependent on the current IoContext (or lack thereof).
+//   3. The BundleDirectoryDelegate (see bundle-fs.h and bundle-fs.c++) is a
+//      read-only directory implementation that is backed by the worker bundle
+//      configuration.
+//
+// The DirectoryDelegate exposes a subset of the kj::Directory API that covers
+// the functionality we support.
+class DirectoryDelegate {
+ public:
+  virtual ~DirectoryDelegate() noexcept(false) = default;
+
+  virtual kj::Own<const DirectoryDelegate> clone() const = 0;
+
+  virtual kj::FsNode::Metadata stat() const {
+    return kj::FsNode::Metadata{
+      .type = kj::FsNode::Type::DIRECTORY,
+      .size = 0,
+      .spaceUsed = 0,
+      .lastModified = kj::UNIX_EPOCH,
+      .linkCount = 1,
+      .hashCode = 0,
+    };
+  }
+
+  virtual bool exists(kj::PathPtr path) const = 0;
+
+  virtual kj::Array<kj::String> listNames() const = 0;
```

Comment:
`listNames` seems redundant and not significantly more efficient than `listEntries`

Reply:
Agreed but this is modeled directly after kj/filesystem which separates these. Since our implementation puts a `kj::ReadableDirectory` behind this API it's better to remain consistent.